app/src/main/java/com/nutomic/syncthingandroid/service/RunConditionMonitor.java

```diff
@@ -134,64 +156,107 @@ public void updateShouldRunDecision() {
             }
         }
 
-        // Always run in background
-        if (!prefAlwaysRunInBackground) {
-            /**
-             * User did not specify run conditions in the options.
-             * The app is displaying a foreground activity and syncthing should run.
-             */
-            Log.v(TAG, "decideShouldRun: !prefAlwaysRunInBackground");
-            return true;
+        // Android global AutoSync setting.
+        if (prefRespectMasterSync && !ContentResolver.getMasterSyncAutomatically()) {
+            Log.v(TAG, "decideShouldRun: prefRespectMasterSync && !getMasterSyncAutomatically");
+            return false;
         }
 
-        // Run only when charging.
-        if (prefRunOnlyWhenCharging && !isCharging()) {
-            Log.v(TAG, "decideShouldRun: prefRunOnlyWhenCharging && !isCharging");
-            return false;
+        // Run on mobile data.
+        if (prefRunOnMobileData && isMobileDataConnection()) {
+            Log.v(TAG, "decideShouldRun: prefRunOnMobileData && isMobileDataConnection");
+            return true;
         }
 
-        // Run only on wifi.
-        if (prefRunOnlyOnWifi && !isWifiOrEthernetConnection()) {
-            // Not on wifi.
-            Log.v(TAG, "decideShouldRun: prefRunOnlyOnWifi && !isWifiOrEthernetConnection");
-            return false;
+        // Run on wifi.
+        if (prefRunOnWifi && isWifiOrEthernetConnection()) {
+            if (prefRunOnMeteredWifi) {
+                // Check if we are on metered wifi and if wifi whitelist run condition is met.
+                if (isMeteredNetworkConnection() && wifiWhitelistConditionMet(prefWifiWhitelistEnabled, whitelistedWifiSsids)) {
+                    Log.v(TAG, "decideShouldRun: prefRunOnWifi && isWifiOrEthernetConnection && prefRunOnMeteredWifi && isMeteredNetworkConnection && wifiWhitelistConditionMet");
+                    return true;
+                }
+            } else {
+                // Check if we are on a non-metered wifi and if wifi whitelist run condition is met.
+                if (!isMeteredNetworkConnection() && wifiWhitelistConditionMet(prefWifiWhitelistEnabled, whitelistedWifiSsids)) {
+                    Log.v(TAG, "decideShouldRun: prefRunOnWifi && isWifiOrEthernetConnection && !prefRunOnMeteredWifi && !isMeteredNetworkConnection && wifiWhitelistConditionMet");
+                    return true;
+                }
+            }
         }
 
-        // Run only on whitelisted wifi ssids.
-        if (prefRunOnlyOnWifi && prefWifiWhitelistEnabled) {
-            // Wifi connection detected. Wifi ssid whitelist enabled.
-            Log.v(TAG, "decideShouldRun: prefRunOnlyOnWifi && prefWifiWhitelistEnabled");
-            return isWifiConnectionWhitelisted(whitelistedWifiSsids);
+        // Run in flight mode.
+        if (prefRunInFlightMode && isFlightMode()) {
+            Log.v(TAG, "decideShouldRun: prefRunInFlightMode && isFlightMode");
+            return true;
```

Comment:
After second look, flight mode doesn't actually check for flight mode, but rather for active connection, in which case this function `isFlightMode` should be renamed to represent what it actually does, ala `isConnectedToAnyNetwork`

Reply:
isFlightMode is a function that works differently on Android 4.x than on 5+ phones. It's a little bit tricky on Android 5+, for example. To explain it on your example.
( If I am connected to a metered wifi, and prefer not to run over metered wifi, yet happen to be in flight mode ) ... you'll get an isFlightMode == false so syncthing will not run.

This is why the name is correct - even if the content of the function (as you've observed right) does something looking completely different from flight mode detection. We have this ni == null "hacky" part in since "@Nutomic" found it as its the only one compatible down until Android 4.x In newer android versions, the ConnectivityManager got expanded in android and can return a connection type flight mode. I see no reason to change the function name as it only will return true if we (really) are in flight mode. I've also tested this. What buggy phones will do is written on the paper of future issues.